File: src/org/gga/graph/impl/EdgeMapImpl.java
package org.gga.graph.impl;

import org.gga.graph.maps.EdgeMap;
import org.gga.graph.Edge;

/**
 * @author mike
 */
public class EdgeMapImpl<E> implements EdgeMap<E> {
    private Object[] data = new Object[10];

    public void put(Edge e, E data) {
        int idx = e.idx();

        if (idx >= this.data.length) {
            Object[] newData = new Object[Math.max(this.data.length * 2, idx + 1)];
            System.arraycopy(this.data, 0, newData, 0, this.data.length);
            this.data = newData;
        }

        this.data[idx] = data;
    }

    @SuppressWarnings({"unchecked"})
    public E get(Edge e) {
        return (E) data[e.idx()];
    }
}

File: src/main/scala/org/gga/graph/impl/EdgeMapImpl.scala
package org.gga.graph.impl

import org.gga.graph.Edge
import org.gga.graph.maps.EdgeMap
import javax.annotation.Nullable
import scala.None
import collection.mutable

/**
 * @author mike
 */
class EdgeMapImpl[E] extends EdgeMap[E] {
  private var data: mutable.ArraySeq[E] = mutable.ArraySeq.empty

  def put(e: Edge, data: E) {
    val idx: Int = e.getIdx
    if (idx >= this.data.length) {
      val newLen: Int = idx max (this.data.length + 1) * 2
      this.data = this.data.padTo(newLen, null.asInstanceOf[E])
    }
    this.data(idx) = data
  }

  @Nullable
  def get(e: Edge): Option[E] = {
    val idx: Int = e.getIdx
    if (idx >= data.length) return None
    Some(data(idx))
  }
}